### PR TITLE
Add StepLR scheduler

### PR DESCRIPTION
**Pull request by gitttt-1234**
Currently, we support [`ReduceLROnPlateau`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html) for reducing learning rate by monitoring the `val_loss`. In this PR, we add support for [`StepLR`](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html), to have a fixed learning rate schedule (for eg, reducing lr by 0.5 every 30 epochs).

## Summary by CodeRabbit

- **New Features**
	- Enhanced configuration options for learning rate scheduling and early stopping mechanisms.
	- Introduced new parameters for data processing, including `chunk_size` and `user_instances_only`.
	- Added structured configuration for learning rate adjustments using `ReduceLROnPlateau`.

- **Bug Fixes**
	- Improved error handling for unsupported learning rate schedulers in the training process.

- **Documentation**
	- Updated configuration documentation to reflect new parameters and improved clarity.

- **Tests**
	- Expanded test suite for `ModelTrainer` to validate new learning rate scheduler configurations and error handling.

### CODE REVIEW

File: docs/config.md [hunk 1]
```diff
@@ -176,13 +176,17 @@ The config file has three main sections:
...
-        - `threshold_mode`: (str) One of "rel", "abs". In rel mode, dynamic_threshold = best * ( 1 + threshold ) in max mode or best * ( 1 - threshold ) in min mode. In abs mode, dynamic_threshold = best + threshold in max mode or best - threshold in min mode. *Default*: "rel".
-        - `cooldown`: (int) Number of epochs to wait before resuming normal operation after lr has been reduced. *Default*: 0
-        - `patience`: (int) Number of epochs with no improvement after which learning rate will be reduced. For example, if patience = 2, then we will ignore the first 2 epochs with no improvement, and will only decrease the LR after the third epoch if the loss still hasn’t improved then. *Default*: 10.
-        - `factor`: (float) Factor by which the learning rate will be reduced. new_lr = lr * factor. *Default*: 0.1.
-        - `min_lr`: (float or List[float]) A scalar or a list of scalars. A lower bound on the learning rate of all param groups or each group respectively. *Default*: 0.
+        - `use_step_lr`: (bool) True if `torch.optim.lr_scheduler.StepLR` should be used, else `torch.optim.lr_scheduler.ReduceLROnPlateau` will be used.
```
Comment (talmo) [inline docs/config.md:179]: We should probably make this more general so we can support different learning rate schedulers.

How about:

```suggestion
        - `lr_scheduler`: (str) Name of the scheduler to use. Valid schedulers: `"StepLR"`, `"ReduceLROnPlateau"`
```

File: docs/config_bottomup.yaml [hunk 2]
```diff
@@ -104,12 +105,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    threshold_mode: abs
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline docs/config_bottomup.yaml:108]: see above about LR scheduler parametrization

File: docs/config_centroid.yaml [hunk 3]
```diff
@@ -133,12 +134,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    threshold_mode: abs
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline docs/config_centroid.yaml:137]: see above about LR scheduler parametrization

File: docs/config_topdown_centered_instance.yaml [hunk 4]
```diff
@@ -102,12 +104,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    threshold_mode: abs
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline docs/config_topdown_centered_instance.yaml:107]: see above about LR scheduler parametrization

File: sleap_nn/training/model_trainer.py [hunk 5]
```diff
@@ -636,16 +636,25 @@ def configure_optimizers(self):
...
-            min_lr=self.trainer_config.lr_scheduler.min_lr,
+            threshold=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.threshold,
+            threshold_mode=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.threshold_mode,
+            cooldown=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.cooldown,
+            patience=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.patience,
+            factor=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.factor,
+            min_lr=self.trainer_config.lr_scheduler.reduce_lr_on_plateau.min_lr,
         )
+
+        if self.trainer_config.lr_scheduler.use_step_lr:
```
Comment (talmo) [inline sleap_nn/training/model_trainer.py:651]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance/initial_config.yaml [hunk 6]
```diff
@@ -82,11 +86,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline tests/assets/minimal_instance/initial_config.yaml:89]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance/training_config.yaml [hunk 7]
```diff
@@ -95,11 +99,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: false
```
Comment (talmo) [inline tests/assets/minimal_instance/training_config.yaml:102]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance_bottomup/initial_config.yaml [hunk 8]
```diff
@@ -84,11 +87,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline tests/assets/minimal_instance_bottomup/initial_config.yaml:90]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance_bottomup/training_config.yaml [hunk 9]
```diff
@@ -99,11 +102,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline tests/assets/minimal_instance_bottomup/training_config.yaml:105]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance_centroid/initial_config.yaml [hunk 10]
```diff
@@ -78,11 +81,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline tests/assets/minimal_instance_centroid/initial_config.yaml:84]: see above about LR scheduler parametrization

File: tests/assets/minimal_instance_centroid/training_config.yaml [hunk 11]
```diff
@@ -90,11 +93,14 @@ trainer_config:
     lr: 0.0001
     amsgrad: false
   lr_scheduler:
-    threshold: 1.0e-07
-    cooldown: 3
-    patience: 5
-    factor: 0.5
-    min_lr: 1.0e-08
+    use_step_lr: False
```
Comment (talmo) [inline tests/assets/minimal_instance_centroid/training_config.yaml:96]: see above about LR scheduler parametrization

File: tests/fixtures/datasets.py [hunk 12]
```diff
@@ -153,12 +156,15 @@ def config(sleap_data_dir):
                 "optimizer_name": "Adam",
                 "optimizer": {"lr": 0.0001, "amsgrad": False},
                 "lr_scheduler": {
-                    "threshold": 1e-07,
-                    "threshold_mode": "rel",
-                    "cooldown": 3,
-                    "patience": 5,
-                    "factor": 0.5,
-                    "min_lr": 1e-08,
+                    "use_step_lr": False,
```
Comment (talmo) [inline tests/fixtures/datasets.py:159]: see above about LR scheduler parametrization

File: tests/training/test_model_trainer.py [hunk 13]
```diff
@@ -99,11 +100,14 @@ def test_trainer(config, tmp_path: str, minimal_instance_bottomup_ckpt: str):
 
     #######
 
-    # update save_ckpt to True
+    # update save_ckpt to True and test step lr
     OmegaConf.update(config, "trainer_config.save_ckpt", True)
     OmegaConf.update(config, "trainer_config.use_wandb", True)
     OmegaConf.update(config, "data_config.preprocessing.crop_hw", None)
     OmegaConf.update(config, "data_config.preprocessing.min_crop_size", 100)
+    OmegaConf.update(config, "trainer_config.lr_scheduler.use_step_lr", True)
```
Comment (talmo) [inline tests/training/test_model_trainer.py:108]: see above about LR scheduler parametrization

**Review by talmo — CHANGES_REQUESTED** [review]

**Review by talmo — APPROVED** [review]
